Error when using validation for sub-form.

The validation of sub-form in Play framework cause internal errors.

### Play Version
2.8.2


### API
Java


### Operating System
Windows 10 and ArchLinux


### JDK
11.0.8.10 and 1.8.0.265

### Library Dependencies
None

### Expected Behavior

When having the following code : 

``` java
@Constraints.Validate
public class SubForm implements Constraints.Validatable<List<ValidationError>> {
    @Constraints.Required
    protected String name;

    @Constraints.Required
    protected Integer value;

    // Getters, Setters

    @Override
    public List<ValidationError> validate() {
        List<ValidationError> errors = new ArrayList();
        // Error for test purpose.
        errors.add(new ValidationError("value", "Invalid value"));
        return errors.isEmpty() ? null : errors;
    }
}
```

``` java
import javax.validation.Valid;

@Constraints.Validate
public class MainForm implements Constraints.Validatable<List<ValidationError>> {
    @Constraints.Required
    protected String foo;

    @Constraints.Required
    @Valid
    protected List<SubForm> entries;

    // Getters, Setters

    @Override
    public List<ValidationError> validate() {
        List<ValidationError> errors = new ArrayList();
        // Validation but let's say it has no error.
        return errors.isEmpty() ? null : errors;
    }
}
```

The controller
``` java
public class MyController extends AController {
    private final FormFactory formFactory;

    @Inject
    public MyController(final MessagesApi messagesApi, final FormFactory formFactory) {
        this.formFactory = formFactory;
    }

    public Result POST_SubmitForm(final Http.Request request) {
        final Form<MainForm> boundForm = this.formFactory.form(MainForm.class).bindFromRequest(request);

        if (boundForm.hasErrors()) {
            return Results.badRequest(boundForm.errorsAsJson());
        }
        // Logic if the form is ok.
        return Results.noContent();
    }
}
```

So ... With that code, what happens is that when the route is called, the `MainForm.validate` method is called (which is normal). The `@Valid` on the list of subForm make so that the `@Constraints.Required` are called (normal too) and call the `SubForm.validate` method which is expected to be able to add custom errors on the form.


### Actual Behavior

The validation crash because the errors added on the `SubForm.validate` seems to be applied to the `MainForm`.
Here is the stacktrace.

```
play.api.http.HttpErrorHandlerExceptions$$anon$1: Execution exception[[IllegalStateException: JSR-303 validated property 'entries[2]' does not have a corresponding accessor for data binding - check your DataBinder's configuration (bean property versus direct field access)]]
	at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:359)
	at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:261)
	at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:429)
	at play.core.server.AkkaHttpServer$$anonfun$2.applyOrElse(AkkaHttpServer.scala:421)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:453)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:56)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:93)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:93)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:48)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.IllegalStateException: JSR-303 validated property 'entries[2]' does not have a corresponding accessor for data binding - check your DataBinder's configuration (bean property versus direct field access)
	at play.data.Form.addConstraintViolationToBindingResult(Form.java:929)
	at play.data.Form.lambda$bind$10(Form.java:1028)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at play.data.Form.bind(Form.java:1028)
	at play.data.Form.bindFromRequest(Form.java:665)
	at controllers.api.MyController.POST_SubmitForm(MyController.java:46)
	at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$14(Routes.scala:199)
	at play.core.routing.HandlerInvokerFactory$$anon$8.resultCall(HandlerInvoker.scala:150)
	at play.core.routing.HandlerInvokerFactory$$anon$8.resultCall(HandlerInvoker.scala:149)
	at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:115)
	at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:119)
	at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:33)
	at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:175)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:671)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:430)
	at play.core.j.HttpExecutionContext.$anonfun$execute$1(HttpExecutionContext.scala:64)
	at play.api.libs.streams.Execution$trampoline$.execute(Execution.scala:70)
	at play.core.j.HttpExecutionContext.execute(HttpExecutionContext.scala:59)
	at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:392)
	at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:302)
	at scala.concurrent.impl.Promise$DefaultPromise.dispatchOrAddCallbacks(Promise.scala:276)
	at scala.concurrent.impl.Promise$DefaultPromise.map(Promise.scala:146)
	at scala.concurrent.Future$.apply(Future.scala:671)
	at play.core.j.JavaAction.apply(JavaAction.scala:176)
	at play.api.mvc.Action.$anonfun$apply$4(Action.scala:82)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:433)
	... 12 common frames omitted
Caused by: org.springframework.beans.NotReadablePropertyException: Invalid property 'value' of bean class [forms.api.MainForm]: Bean property 'value' is not readable or has an invalid getter method: Does the return type of the getter match the parameter type of the setter?
	at org.springframework.beans.AbstractNestablePropertyAccessor.getPropertyValue(AbstractNestablePropertyAccessor.java:622)
	at org.springframework.beans.AbstractNestablePropertyAccessor.getPropertyValue(AbstractNestablePropertyAccessor.java:612)
	at org.springframework.validation.AbstractPropertyBindingResult.getActualFieldValue(AbstractPropertyBindingResult.java:104)
	at org.springframework.validation.AbstractBindingResult.rejectValue(AbstractBindingResult.java:117)
	at play.data.Form.lambda$addConstraintViolationToBindingResult$7(Form.java:912)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at play.data.Form.addConstraintViolationToBindingResult(Form.java:910)
	... 37 common frames omitted
```

--- a/web/play-java-forms/src/main/java/play/data/Form.java
+++ b/web/play-java-forms/src/main/java/play/data/Form.java
@@ -934,25 +934,10 @@ public class Form<T> {
               (String) dynamicPayload // dynamicPayload itself is the error message(-key)
               );
         } else if (dynamicPayload instanceof ValidationError) {
-          final ValidationError error = (ValidationError) dynamicPayload;
-          result.rejectValue(
-              error.key(),
-              violation.getConstraintDescriptor().getAnnotation().annotationType().getSimpleName(),
-              error.arguments() != null ? error.arguments().toArray() : new Object[0],
-              error.message());
+          rejectValidationError(violation, result, (ValidationError) dynamicPayload, field);
         } else if (dynamicPayload instanceof List) {
           ((List<ValidationError>) dynamicPayload)
-              .forEach(
-                  error ->
-                      result.rejectValue(
-                          error.key(),
-                          violation
-                              .getConstraintDescriptor()
-                              .getAnnotation()
-                              .annotationType()
-                              .getSimpleName(),
-                          error.arguments() != null ? error.arguments().toArray() : new Object[0],
-                          error.message()));
+              .forEach(error -> rejectValidationError(violation, result, error, field));
         } else {
           result.rejectValue(
               field,
@@ -969,6 +954,19 @@ public class Form<T> {
             ex);
       }
     }
+  }
+
+  private static void rejectValidationError(
+      ConstraintViolation<Object> violation,
+      BindingResult result,
+      final ValidationError error,
+      final String field) {
+    final String keyPrefix = (field == null || field.isEmpty() ? "" : field + ".");
+    result.rejectValue(
+        error.key() != null && !error.key().isEmpty() ? keyPrefix + error.key() : error.key(),
+        violation.getConstraintDescriptor().getAnnotation().annotationType().getSimpleName(),
+        error.arguments() != null ? error.arguments().toArray() : new Object[0],
+        error.message());
   }
 
   private List<ValidationError> getFieldErrorsAsValidationErrors(Lang lang, BindingResult result) {

--- a/web/play-java-forms/src/test/scala/play/data/FormSpec.scala
+++ b/web/play-java-forms/src/test/scala/play/data/FormSpec.scala
@@ -11,6 +11,7 @@ import java.util.Optional
 import java.time.LocalDate
 import java.time.ZoneId
 
+import javax.validation.Valid
 import javax.validation.Validation
 import javax.validation.ValidatorFactory
 import javax.validation.{ Configuration => vConfiguration }
@@ -26,6 +27,7 @@ import play.api.inject.guice.GuiceApplicationBuilder
 import play.api.test.WithApplication
 import play.api.Application
 import play.components.TemporaryFileComponents
+import play.data.validation.Constraints
 import play.data.validation.ValidationError
 import play.i18n.Lang
 import play.libs.F
@@ -41,6 +43,7 @@ import play.routing.Router
 import play.test.Helpers
 import play.twirl.api.Html
 
+import javax.validation.constraints.Size
 import scala.beans.BeanProperty
 import scala.collection.JavaConverters._
 import scala.compat.java8.OptionConverters._
@@ -1244,6 +1247,71 @@ trait FormSpec extends CommonFormSpec {
         myForm.errors().size() must beEqualTo(0)
         myForm.errors("name").size() must beEqualTo(0)
       }
+      "when it is located in a subform and but it doesn't get called because subform was not submitted" in {
+        val myForm = formFactory
+          .form(classOf[JavaMainForm])
+          .bind(Lang.defaultLang(), TypedMap.empty(), Map.empty[String, String].asJava)
+        myForm.globalErrors().size() must beEqualTo(0)
+        myForm.errors().size() must beEqualTo(3)
+        myForm.errors("entries").size() must beEqualTo(1)
+        myForm.errors("entries").get(0).message() must beEqualTo("error.required")
+        myForm.errors("entry").size() must beEqualTo(2)
+        myForm.errors("entry").asScala.map(_.message()) must contain("error.required")
+        myForm.errors("entry").asScala.map(_.message()) must contain("validate of parent: I always get called!")
+      }
+      "when it is located in a subform (and sub-subform) and returns an error it should automatically prefix the error key with the parent form field" in {
+        val myForm = formFactory
+          .form(classOf[JavaMainForm])
+          .bind(
+            Lang.defaultLang(),
+            TypedMap.empty(),
+            Map(
+              "entry.name" -> "Bill",
+              //"entry.value" -> "...",  -> Missing but required by validate method of sub form
+              //"entries[0].name"  -> "...", -> Missing but required by @Constraints.Required
+              "entries[0].value" -> "14",
+              "entries[1].name"  -> "John",
+              //"entries[1].value" -> "...",  -> Missing but required by validate method of sub form
+              "entries[0].entries[0].name"   -> "Robin Hood",
+              "entries[0].entries[1].street" -> "Wall Street",
+            ).asJava
+          )
+        myForm.globalErrors().size() must beEqualTo(0)
+        myForm.errors().size() must beEqualTo(9)
+        myForm.errors("entry").size() must beEqualTo(1)
+        myForm.errors("entry").get(0).message() must beEqualTo("validate of parent: I always get called!")
+        myForm.errors("entry.value").size() must beEqualTo(1) // prefixed by Play
+        myForm.errors("entry.value").get(0).message() must beEqualTo("validate of child: value can't be null!")
+        myForm.errors("entries").size() must beEqualTo(1)
+        myForm.errors("entries").get(0).message() must beEqualTo("size must be between 0 and 1")
+        myForm.errors("entries[0].name").size() must beEqualTo(1)
+        myForm.errors("entries[0].name").get(0).message() must beEqualTo("error.required")
+        myForm.errors("entries[1].value").size() must beEqualTo(1) // prefixed by Play
+        myForm.errors("entries[1].value").get(0).message() must beEqualTo("validate of child: value can't be null!")
+        myForm.errors("entries[0].entries[0].value").size() must beEqualTo(1) // prefixed by Play
+        myForm.errors("entries[0].entries[0].value").get(0).message() must beEqualTo(
+          "validate of child of child: value can't be null!"
+        )
+        myForm.errors("entries[0].entries[0].street").size() must beEqualTo(1)
+        myForm.errors("entries[0].entries[0].street").get(0).message() must beEqualTo("error.required")
+        myForm.errors("entries[0].entries[1].name").size() must beEqualTo(1)
+        myForm.errors("entries[0].entries[1].name").get(0).message() must beEqualTo("error.required")
+        myForm.errors("entries[0].entries[1].value").size() must beEqualTo(1) // prefixed by Play
+        myForm.errors("entries[0].entries[1].value").get(0).message() must beEqualTo(
+          "validate of child of child: value can't be null!"
+        )
+      }
+      "when it is located in a form that is sometimes used as sub form but not now and returns an error it should NOT automatically prefix the error key" in {
+        val myForm = formFactory
+          .form(classOf[JavaChildForm])
+          .bind(Lang.defaultLang(), TypedMap.empty(), Map.empty[String, String].asJava)
+        myForm.globalErrors().size() must beEqualTo(0)
+        myForm.errors().size() must beEqualTo(2)
+        myForm.errors("name").size() must beEqualTo(1)
+        myForm.errors("name").get(0).message() must beEqualTo("error.required")
+        myForm.errors("value").size() must beEqualTo(1) // not prefixed by Play
+        myForm.errors("value").get(0).message() must beEqualTo("validate of child: value can't be null!")
+      }
     }
 
     "not process it's legacy validate method when the Validatable interface is implemented" in {
@@ -1321,4 +1389,57 @@ class JavaForm(@BeanProperty var foo: java.util.List[JavaSubForm]) {
 }
 class JavaSubForm(@BeanProperty var a: String, @BeanProperty var b: String) {
   def this() = this(null, null)
+}
+
+@Constraints.Validate
+class JavaMainForm extends Constraints.Validatable[ValidationError] {
+
+  @BeanProperty
+  @Constraints.Required
+  @Size(max = 1)
+  @Valid
+  var entries: java.util.List[JavaChildForm] = _
+
+  @BeanProperty
+  @Constraints.Required
+  @Valid
+  var entry: JavaChildForm = _
+
+  override def validate = new ValidationError("entry", "validate of parent: I always get called!")
+}
+
+@Constraints.Validate
+class JavaChildForm extends Constraints.Validatable[ValidationError] {
+
+  @BeanProperty
+  @Constraints.Required
+  var name: String = _
+
+  @BeanProperty
+  var value: java.lang.Integer = _
+
+  @BeanProperty
+  @Valid
+  var entries: java.util.List[JavaChildChildForm] = _
+
+  override def validate: ValidationError =
+    if (value == null) new ValidationError("value", "validate of child: value can't be null!") else null
+}
+
+@Constraints.Validate
+class JavaChildChildForm extends Constraints.Validatable[ValidationError] {
+
+  @BeanProperty
+  @Constraints.Required
+  var name: String = _
+
+  @BeanProperty
+  @Constraints.Required
+  var street: String = _
+
+  @BeanProperty
+  var value: java.lang.Integer = _
+
+  override def validate: ValidationError =
+    if (value == null) new ValidationError("value", "validate of child of child: value can't be null!") else null
 }

